New module: cgroup

Dayde opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
New module for cgroup configuration deployment
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
cgroup

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.4.3.0
  config file = None
  configured module search path = [u'/Users/dayde/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.10 (default, Jul 15 2017, 17:16:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
In the current state, the module handles only memory limit and memory swap limit. However if you find this useful I'd be glad to improve it!
I'll also add tests if it is to be merged ;-) I would have loved to write them while coding but I couldn't managed to make them run, first time coding with python over here ^^'
<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Dayde committed:
New module: cgroup

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+version_added: "2.6"
+
+description:
+    - Manages cgroup configurations in the /etc/cgconfig.d directory.
```
Use [formatting function](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#formatting-functions) `C()` for path and program name.

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+def run_module():
+    module_args = dict(
+        name=dict(type='str', required=True),
+        state=dict(type='str', required=True),
```
`state`: a `choice` parameter should be used ([example](https://github.com/ansible/ansible/blob/6cdf91fd35fefec35dca4594db231cc6af5a8620/lib/ansible/modules/system/capabilities.py#L167))

Dayde replied:
Done

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+        choices: ['absent', 'present']
+    memlimit:
+        description:
+            - The maximum RSS memory for the cgroup.
```
`memlimit` and `memswaplimit`: allowed units could be mentioned.

Dayde replied:
Done

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+    module.exit_json(**result)
+
+
+def read_config_from_file(desired_cgroup):
```
`read_config_from_file` could return the content of the file or `None` if the file doesn't exist (instead of a tuple).

Dayde replied:
Done

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+            file_content = fh.read()
+            fh.close()
+        return (True, file_content)
+    except:
```
Don't use `except` without exception type ([ref1](https://docs.python.org/2/howto/doanddont.html#except), [ref2](https://docs.quantifiedcode.com/python-anti-patterns/correctness/no_exception_type_specified.html)). Here checking file existence could replace `try/except`.

Dayde replied:
Done

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+
+def read_config_from_file(desired_cgroup):
+    try:
+        with open("/etc/cgconfig.d/%s.conf" % (desired_cgroup), 'r') as fh:
```
Should not the cgroup existence checked there ?

Dayde replied:
Done

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+
+    config = {}
+    iterraw = iter(raw_config.split('\n'))
+    # Skip first line ("# Ansible : <name>")
```
What if the first line is missing ? The file might not have been created by this module.

Dayde replied:
Will now skip any comment line if needed

pilou- commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,256 @@
+    # Skip first line ("# Ansible : <name>")
+    line = next(iterraw)
+    # Skip second line ("group <name> {")
+    line = next(iterraw)
```
The cgroup name must be checked.

Dayde replied:
Done

pilou- reviewed (commented):
Because this module manage content of a file, should not that be a role instead of a module ?

Dayde committed:
Module cgroup: better config parsing

Dayde commented:
@pilou- as you pointed out the config parsing was not very satisfying. The new version is more flexible I think and should do the job. Any comments appreciated :)

resmo commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,297 @@
+    state:
+        description:
+            - Whether the cgroup should be present or absent.
+        required: true
```
by convention the state should be defaulted to `present` whenever possible, as a result th

~~~diff
-        required: true
+        default: present
~~~

Dayde replied:
Done

resmo commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,297 @@
+def run_module():
+    module_args = dict(
+        name=dict(type='str', required=True),
+        state=dict(type='str', required=True, choices=['absent', 'present']),
```
~~~diff
-        state=dict(type='str', required=True, choices=['absent', 'present']),
+        state=dict(type='str', default='present', choices=['absent', 'present']),
~~~

Dayde replied:
Done

Dayde committed:
Module cgroup: PR reviews

Switching state from required to default present.
Fixing an issue in config parser: will continue as expected if there's a space between param value and semi-colon.

Dayde committed:
Module cgroup: adding permission support for cgroups

Dayde committed:
Module cgroup: use ansible diff

danihodovic commented:
@Dayde what is the status of this PR?

Dayde commented:
@danihodovic well, you tell me 😉
I've taken comments into account and was waiting for other reviews or merge

danihodovic commented:
@pilou?

resmo commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,295 @@
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=False
```
check mode seems to be implemented, right?

resmo commented on lib/ansible/modules/system/cgroup.py:
```diff
@@ -0,0 +1,295 @@
+
+short_description: Manage cgroup configurations
+
+version_added: "2.6"
```
please update to 2.9, we consider to merge

Akasurde commented:
@Dayde Are you still working on this, let us know ? Thanks, needs_info

Dayde commented:
@Akasurde I hope to have some time to update in the next few weeks :)
If there's no update until december feel free to close this I'll reopen if I ever get to work on it again

Dayde committed:
Module cgroup: test parsing logic & fix some comment parsing issue

Dayde committed:
Module cgroup: PR reviews

rebased on 2.10
doc now reports check mode is supported

Akasurde commented:
Please submit a new module request to https://github.com/ansible-collections/community.general. Thanks for your contribution.
